### ISSUE TITLE
Authentication Exception Standardization

### ISSUE THREAD

**Issue by IMikhelson** [ccxt/ccxt#4071]
- OS: Windows 10 Pro
- Programming Language version: Python 3.6
- CCXT version: 1.17.439
- Exchange: Binance, Coinbase Pro, others?
- Method: any private method

When I call a private method with an incorrect API key, secret, or passphrase, I get different behavior on different exchanges.  So far, I have tested Binance and Coinbase Pro.  My findings are below:

**Coinbase Pro**
wrong key: AuthenticationError ("coinbasepro Invalid API Key")
wrong secret: ExchangeError ("coinbasepro invalid signature")
wrong phrase: ExchangeError ("coinbasepro Invalid Passphrase")

**Binance:**
wrong key: AuthenticationError ("binance {"code":-2014,"msg":"API-key format invalid."}")
wrong secret: AuthenticationError("binance {"code":-1022,"msg":"Signature for this request is not valid."}")

_To summarize:_
An incorrect secret throws different exception types on different exchanges.  It seems like they should all be AuthenticationError types (including incorrect passphrases).  Also, it would be nice if the message in the exception was standard.  Perhaps the exchange's exact error message could be one value, and then a standardized CCXT message could be another (something like "incorrect API secret").

**Comment by IMikhelson** [comment]
Is there a chance that the string responses will get standardized, or is that too much overhead?
Thanks!

**Comment by kroitor** [comment]
> Is there a chance that the string responses will get standardized, or is that too much overhead?

We differentiate exception types by message instead, so, if you thought of parsing that text and determining your reaction, depending on the text – that isn't really how we would like this library to be used. We think it's way better if you differentiate by exception type. 

In other words, instead of doing this:

```JavaScript
// pseudocode
try {
    ...
} catch (e) { // same as `except` in python
    message = e.message
    if (message === 'a') {
        ... handling cases ...
    } else if (message === 'b') {
        ... 
    } ...
}
```

We would like you to do something similar to this:

```JavaScript
try {
    ...
} catch (e) { // same as `except` in python
    message = e.message
    if (type (e) === AuthenticationError) {
        ...
    } else if (type (e) === AnotherErrorType) {
        ...
    }
}
```

In the latter case the message is always the string coming from the exchange, with an exchange id prepended to it, but, again, it is not for parsing, but merely for debugging (mostly).

Now, we do understand, that we currently don't recognize and map all errors to corresponding exception types properly, but, again, this lib is a work in progress, and if you keep reporting inconsistencies, we will fix them all sooner or later.

Let me know if that does not answer your question.

**Comment by IMikhelson** [comment]
Thank you for the detailed response.  In my application, I would like to be able to tell between an authentication error caused by an invalid key vs. an invalid secret (vs. an invalid passphrase).  Since they are all authentication errors, there is not currently a way to distinguish them.  Is this something that may get done in the future?

**Comment by kroitor** [comment]
>  Is this something that may get done in the future?

Yes, for sure! We will add subclasses of AuthenticationError, namely, like you want them exactly:
ExchangeError → AuthenticationError → InvalidApiKey
ExchangeError → AuthenticationError → InvalidSecret

We just didn't have a chance to get our hands on that aspect ;)

Your contributions are welcome, by the way: https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#how-to-contribute-code )

**Comment by kroitor** [comment]
@IMikhelson jfyi, errors are declared here:

- https://github.com/ccxt/ccxt/blob/master/js/base/errors.js
- https://github.com/ccxt/ccxt/tree/master/php
- https://github.com/ccxt/ccxt/blob/master/python/ccxt/base/errors.py

**Comment by IMikhelson** [comment]
Thank you!  I don't know if it's helpful, but I ran the following script, and got the results below.

```
import ccxt

def check_api_info(exchange, key, sec, phrase):
    exchange_class = getattr(ccxt, exchange)
    client = exchange_class({
        'apiKey': key,
        'secret': sec,
        'password': phrase,
        'enableRateLimit': True,
        'ratelimit': 100
    })
    
    try:
        client.load_markets()
    except:
        print(id + ": couldn't load markets")

    try:
        bal = client.fetch_balance()
        print(id + ": Success!")
    except ccxt.AuthenticationError as e:
        print(id + ": Auth Error")
    except Exception as e:
        print(id + ": General Error")


x = ccxt.exchanges
for id in x:
    check_api_info(id,'a','a','a')
```

_1broker: couldn't load markets
_1broker: General Error
_1btcxe: General Error
acx: Auth Error
allcoin: couldn't load markets
allcoin: General Error
anxpro: General Error
anybits: Auth Error
bcex: Auth Error
bibox: couldn't load markets
bibox: Auth Error
bigone: General Error
binance: Auth Error
bit2c: Success!
bitbank: Auth Error
bitbay: General Error
bitfinex: Auth Error
bitfinex2: General Error
bitflyer: Auth Error
bitforex: Auth Error
bithumb: General Error
bitibu: couldn't load markets
bitibu: General Error
bitkk: Auth Error
bitlish: General Error
bitmarket: General Error
bitmex: General Error
bitsane: Auth Error
bitso: Auth Error
bitstamp: Auth Error
bitstamp1: Auth Error
bittrex: Auth Error
bitz: Auth Error
bl3p: General Error
bleutrade: Auth Error
braziliex: General Error
btcalpha: Auth Error
btcbox: General Error
btcchina: General Error
btcexchange: General Error
btcmarkets: General Error
btctradeim: General Error
btctradeua: Success!
btcturk: couldn't load markets
btcturk: General Error
btcx: General Error
buda: Auth Error
bxinth: General Error
ccex: General Error
cex: Auth Error
chbtc: General Error
chilebit: General Error
cobinhood: General Error
coinbase: Auth Error
coinbaseprime: General Error
coinbasepro: General Error
coincheck: Auth Error
coinegg: couldn't load markets
coinegg: General Error
coinex: Auth Error
coinexchange: General Error
coinfalcon: General Error
coinfloor: General Error
coingi: couldn't load markets
coingi: General Error
coinmarketcap: General Error
coinmate: Auth Error
coinnest: General Error
coinone: General Error
coinsecure: General Error
coinspot: Success!
cointiger: Auth Error
coolcoin: couldn't load markets
coolcoin: General Error
crypton: General Error
cryptopia: General Error
deribit: General Error
dsx: General Error
ethfinex: Auth Error
exmo: Auth Error
exx: Auth Error
fcoin: Auth Error
flowbtc: couldn't load markets
flowbtc: General Error
foxbit: General Error
fybse: General Error
fybsg: General Error
gatecoin: General Error
gateio: General Error
gdax: General Error
gemini: General Error
getbtc: General Error
hadax: Auth Error
hitbtc: Auth Error
hitbtc2: General Error
huobi: General Error
huobicny: couldn't load markets
huobicny: General Error
huobipro: Auth Error
ice3x: General Error
independentreserve: General Error
indodax: Auth Error
itbit: Auth Error
jubi: General Error
kkex: General Error
kraken: General Error
kucoin: Auth Error
kuna: Auth Error
lakebtc: Auth Error
lbank: Auth Error
liqui: Auth Error
liquid: Auth Error
livecoin: General Error
luno: Auth Error
lykke: General Error
mercado: General Error
mixcoins: General Error
negociecoins: General Error
nova: Auth Error
okcoincny: General Error
okcoinusd: Auth Error
okex: Auth Error
paymium: Auth Error
poloniex: Auth Error
qryptos: Auth Error
quadrigacx: Auth Error
quoinex: Auth Error
rightbtc: General Error
southxchange: General Error
surbitcoin: General Error
theocean: General Error
therock: Auth Error
tidebit: Auth Error
tidex: Auth Error
uex: Auth Error
urdubit: General Error
vaultoro: Auth Error
vbtc: General Error
virwox: General Error
wex: General Error
xbtce: couldn't load markets
xbtce: Auth Error
yobit: General Error
yunbi: couldn't load markets
yunbi: General Error
zaif: General Error
zb: Auth Error

**Comment by kroitor** [comment]
@IMikhelson yep, thanks for the info, we will proceed adding those handlers where missing... As you can see, there's still plenty of room for development and unification, we welcome all PRs and contributions )
